src/main/java/org/graylog/plugins/beats/BeatsCodec.java

```diff
@@ -70,157 +77,58 @@ public Message decode(@Nonnull RawMessage rawMessage) {
 
-    /**
-     * @see <a href="https://www.elastic.co/guide/en/beats/metricbeat/5.1/exported-fields.html">Metricbeat Exported Fields</a>
-     */
-    private Message parseMetricbeat(Map<String, Object> event) {
-        final Message gelfMessage = createMessage("-", event);
-        gelfMessage.addField("facility", "metricbeat");
-        final Map<String, Object> flattened = flatten(event, "metricbeat", MAP_KEY_SEPARATOR);
-
-        // Fix field names containing dots, like "cpu.name"
-        final Map<String, Object> withoutDots = MapUtils.replaceKeyCharacter(flattened, '.', MAP_KEY_SEPARATOR.charAt(0));
-        gelfMessage.addFields(withoutDots);
+        addFlattened(gelfMessage, rootPath, event);
         return gelfMessage;
     }
 
-    /**
-     * @see <a href="https://www.elastic.co/guide/en/beats/packetbeat/1.2/exported-fields.html">Packetbeat Exported Fields</a>
-     */
-    private Message parsePacketbeat(Map<String, Object> event) {
-        final Message gelfMessage = createMessage("-", event);
-        gelfMessage.addField("facility", "packetbeat");
-        final Map<String, Object> flattened = flatten(event, "packetbeat", MAP_KEY_SEPARATOR);
-
-        // Fix field names containing dots, like "icmp.version"
-        final Map<String, Object> withoutDots = MapUtils.replaceKeyCharacter(flattened, '.', MAP_KEY_SEPARATOR.charAt(0));
-        gelfMessage.addFields(withoutDots);
-
-        return gelfMessage;
-    }
-
-    /**
-     * @see <a href="https://www.elastic.co/guide/en/beats/winlogbeat/1.2/exported-fields.html">Winlogbeat Exported Fields</a>
-     */
-    private Message parseWinlogbeat(Map<String, Object> event) {
-        final String message = String.valueOf(event.remove("message"));
-        final Message gelfMessage = createMessage(message, event);
-        gelfMessage.addField("facility", "winlogbeat");
-        final Map<String, Object> flattened = flatten(event, "winlogbeat", MAP_KEY_SEPARATOR);
-
-        // Fix field names containing dots, like "user.name"
-        final Map<String, Object> withoutDots = MapUtils.replaceKeyCharacter(flattened, '.', MAP_KEY_SEPARATOR.charAt(0));
-        gelfMessage.addFields(withoutDots);
-        return gelfMessage;
+    private void addFlattened(Message message, String currentPath, JsonNode jsonNode) {
+        if (jsonNode.isObject()) {
+            final Iterator<Map.Entry<String, JsonNode>> it = jsonNode.fields();
+            final String pathPrefix = currentPath.isEmpty() ? "" : currentPath + MAP_KEY_SEPARATOR;
+            while (it.hasNext()) {
+                final Map.Entry<String, JsonNode> entry = it.next();
+                addFlattened(message, pathPrefix + entry.getKey(), entry.getValue());
+            }
+        } else if (jsonNode.isArray()) {
+            final List<Object> values = new ArrayList<>(jsonNode.size());
+            for (int i = 0; i < jsonNode.size(); i++) {
+                values.add(valueNode(jsonNode.get(i)));
```

Comment:
When looking at the test for `packetbeat-dns.json` I noticed that objects in arrays aren't getting expanded.
Pretty sure the code before didn't do that before, so we might want to push a fix out to a separate issue, but should it flatten those by adding the index to the `currentPath`?

e.g:
```
{
  "@metadata": {
    "beat": "packetbeat"
  },
  "@timestamp": "2016-04-01T00:00:00.000Z",
  "beat": {
    "hostname": "example.local",
    "name": "example.local"
  },
...
  "dns": {
    "additionals_count": 0,
    "answers": [
      {
        "class": "IN",
        "data": "216.58.211.46",
        "name": "www3.l.google.com",
        "ttl": 845,
        "type": "A"
      }
    ],
...
```

would yield fields like:
```
packetbeat_dns_answers_0_class: "IN"
packetbeat_dns_answers_0_name: "www3.l.google.com"
etc
```

Currently the `answers`  field is just: `packetbeat_dns_answers: ""`

Reply:
✅ I added some code to expand objects nested in arrays but emit literal values in arrays as list, so that these can be queried accordingly (i. e. `tags`, see `BeatsCodecTest#decodeMessagesHandlesFilebeatMessages()`).